### ISSUE TITLE
Is it intended that traits cannot be satisfied by union types?

### ISSUE THREAD

**Issue by pkalikman** [mauro3/Traits.jl#33]
E.g.: 

```
using Traits

@traitdef SomeTrait{X} begin
    some_function(X) -> Int64
end

immutable TypeOne
    x::Float64
end

immutable TypeTwo
    x::Int64
end

typealias FloatOrInt Union{TypeOne,TypeTwo}

function some_function(x::FloatOrInt)
    isa(x,Float64) ? round(Int,x) : x
end
```

```
istrait(SomeTrait{FloatOrInt})
ERROR: Traits.TraitException("Trait SomeTrait{Union{TypeOne,TypeTwo}} has no type parameters.")
```

I traced this in the source to `hasparameters`, which seems it rejects this type because the union type is not a `DataType`.

Is there a principled reason for that? Offhand I don't see why this trait shouldn't be satisfied by the example.

Thanks

**Comment by mauro3** [comment]
Yes, I think this should work but is not implemented at the moment. If a trait is fulfilled for all types of a union then the union should be part of the trait too.

**Comment by mauro3** [comment]
I not gonna have time to work on this for a few weeks.  If you want to have a crack at it, probably need to insert a loop at https://github.com/mauro3/Traits.jl/blob/cd4263735c23f864c7a75d740f5c6c29c0d2afff/src/Traits.jl#L143 which loops over all unions and calls istrait recursively.
